samza-core/src/main/java/org/apache/samza/zk/ZkUtils.java

```diff
@@ -65,11 +65,21 @@
   private volatile String ephemeralPath = null;
   private final ZkKeyBuilder keyBuilder;
   private final int connectionTimeoutMs;
+  private int currentGeneration;
 
   public ZkUtils(ZkKeyBuilder zkKeyBuilder, ZkClient zkClient, int connectionTimeoutMs) {
     this.keyBuilder = zkKeyBuilder;
     this.connectionTimeoutMs = connectionTimeoutMs;
     this.zkClient = zkClient;
+    currentGeneration = 0;
+  }
+
+  public synchronized void incGeneration() {
```

Comment:
Over time I've generally found it a best practice not to synchronize on your object directly because other users may end up doing the same which can lead to surprising locking issues. You can use a private Object as a lock.

However, given that there is no real coordination of state here, you could opt to make currentGeneration an AtomicInt and use CAS to update it. Reads should be cheaper too.